Improve error message when name is omitted but generateName is available

derekwaynecarr opened the pull request:
Objects may omit name and just provide generateName.

If an admission controller blocked it today, it caused a poor error message.  

Fixed up to use the GenerateName prefix if name is not available.

/cc @jwforres

derekwaynecarr committed:
Improve error message when name is omitted but generateName is available

jwforres commented:
@derekwaynecarr example of what the output looks like now ?

derekwaynecarr commented:
@jwforres 

```shell
Events:
  FirstSeen				LastSeen			Count	From				SubobjectPath	Reason		Message
  Tue, 14 Apr 2015 17:48:17 -0400	Tue, 14 Apr 2015 18:15:15 -0400	329	{replication-controller }			failedCreate	Error creating: pods "demo-" is forbidden: Limited to 10 pods
```

lavalamp commented:
Everything in this pr is fine but I dispute that we should force every client to understand this api detail. My understanding is that GenerateName causes apiserver to generate a name? If so, apiserver should do that before passing the object through any filters or storing it. I think we will have many fewer bugs if we don't require everyone to check both Name and GenerateName fields...

derekwaynecarr commented:
@lavalamp - I think this is a problem unique to admission controllers which are invoked prior to name generation having occurred in the storage tier.  I am happy to take a lot of this code and put in a shared function in the admission package to alleviate some of the concern?

lavalamp commented:
I see. Is there a reason specifically why we don't want to do the generation step before sending to admission controller?

derekwaynecarr commented:
I can follow up on that question.

I have an update to this PR that centralizes the issue noted around generate name logic in multiple places.

Will ping for review when ready.

Sent from my iPhone

> On Apr 15, 2015, at 1:28 PM, Daniel Smith <notifications@github.com> wrote:
> 
> I see. Is there a reason specifically why we don't want to do the generation step before sending to admission controller?
> 
> —
> Reply to this email directly or view it on GitHub.
>

derekwaynecarr commented:
@lavalamp - please take another pass for review.

I can follow-up on moving generateName behavior in a subsequent PR if its a major concern, but right now, I localized all the logic into admission.NewForbidden.

A few other major changes in this defect worth noting:

1. Admission controllers now only throw api/errors back to the REST layer.
2. Kind is now made available to admission control (per discussion with @smarterclayton )

For illustration of new error message example (see failedCreate events):

```shell
cluster/kubectl.sh describe rc demo --namespace=test
current-context: "vagrant"
Running: cluster/../cluster/vagrant/../../_output/dockerized/bin/linux/amd64/kubectl --kubeconfig=/home/decarr/.kubernetes_vagrant_kubeconfig describe rc demo --namespace=test
W0415 23:40:30.657741   25946 request.go:289] field selector: v1beta1 - events - involvedObject.uid - 469d0aff-e3ea-11e4-a8cf-0800279696e1: need to check if this is versioned correctly.
W0415 23:40:30.657807   25946 request.go:289] field selector: v1beta1 - events - involvedObject.id - demo: need to check if this is versioned correctly.
W0415 23:40:30.657816   25946 request.go:289] field selector: v1beta1 - events - involvedObject.namespace - test: need to check if this is versioned correctly.
W0415 23:40:30.657825   25946 request.go:289] field selector: v1beta1 - events - involvedObject.kind - ReplicationController: need to check if this is versioned correctly.
Name:		demo
Image(s):	openshift/hello-openshift
Selector:	run-container=demo
Labels:		run-container=demo
Replicas:	10 current / 11 desired
Pods Status:	0 Running / 10 Waiting / 0 Succeeded / 0 Failed
Events:
  FirstSeen				LastSeen			Count	From				SubobjectPath	Reason		Message
  Wed, 15 Apr 2015 23:40:07 -0400	Wed, 15 Apr 2015 23:40:20 -0400	55	{replication-controller }			failedCreate	Error creating: Pod "demo-" is forbidden: Unable to CREATE pods at this time because there was an error enforcing quota
  Wed, 15 Apr 2015 23:40:20 -0400	Wed, 15 Apr 2015 23:40:27 -0400	4	{replication-controller }			failedCreate	Error creating: Pod "demo-" is forbidden: Limited to 10 pods
```

smarterclayton commented:
Generate name was in the storage layer because the storage layer may be able to generate names more efficiently than the api server.   While error messages are possibly more complicated, I'm not sure I agree that's a better factoring to do it earlier.

Also, when we have to retry create because he name is taken, I want it as close to the store as possible.

smarterclayton commented on pkg/api/meta/interfaces.go:
```diff
@@ -40,6 +40,8 @@ type Interface interface {
 	SetNamespace(namespace string)
 	Name() string
 	SetName(name string)
+	GenerateName() string
```
I think at this point we should be using api.ObjectMetaFor() instead of metadata.Accessor.  Most of the meta abstraction is unnecessary (and much less efficient).

lavalamp replied:
+1 the accessor made more sense back when I added it; it's pretty heavyweight now.

lavalamp commented:
>Generate name was in the storage layer because the storage layer may be able to generate names more efficiently than the api server. While error messages are possibly more complicated, I'm not sure I agree that's a better factoring to do it earlier.

OK, that makes sense. I'll rescind my request.

lavalamp commented on pkg/admission/errors.go:
```diff
@@ -0,0 +1,36 @@
+	if obj != nil {
+		name, _ = meta.NewAccessor().Name(obj)
+		if len(name) == 0 {
+			name, _ = meta.NewAccessor().GenerateName(obj)
```
perhaps go back to unknown if both are empty? Or does that not pass validation and hence never get here?

lavalamp commented on pkg/admission/errors.go:
```diff
@@ -0,0 +1,36 @@
+	obj := a.GetObject()
+	if obj != nil {
+		name, _ = meta.NewAccessor().Name(obj)
+		if len(name) == 0 {
```
Perhaps comment that this is only necessary because the name generation hasn't been done when admission runs.

lavalamp commented on plugin/pkg/admission/limitranger/admission.go:
```diff
@@ -179,6 +181,8 @@ func mergePodResourceRequirements(pod *api.Pod, defaultRequirements *api.Resourc
 // requirements if not specified, and enumerated on the LimitRange
 func PodLimitFunc(limitRange *api.LimitRange, pod *api.Pod) error {
 
+	fmt.Printf("POD %v\n", pod)
```
debugging cruft?

lavalamp commented on pkg/admission/errors.go:
```diff
@@ -0,0 +1,36 @@
+)
+
+// NewForbidden is a utility function to return a well-formatted admission control error response
+func NewForbidden(a Attributes, internalError error) error {
```
I recommend checking the type of internalError; specifically, if it's already an apierror of type Forbidden, don't double wrap it.

lavalamp replied:
(if there was an easy way to tell if you were running in a test I'd ask for it to panic in that case, but that's hard)

lavalamp commented on pkg/admission/attributes.go:
```diff
@@ -21,14 +21,16 @@ import (
 }
 
-func NewAttributesRecord(object runtime.Object, namespace, resource, operation string) Attributes {
+func NewAttributesRecord(object runtime.Object, kind, namespace, resource, operation string) Attributes {
```
kind is deducible from resource and vs versa-- do you really need to make everyone pass both?

derekwaynecarr replied:
I think I can make an argument to provide Resource, Kind, and APIVersion in the long run.

Since the AttributesRecord is only constructed in the 4 places, I don't think the extra argument is a major concern.  The Attributes interface makes the consumer side easy enough I think.

lavalamp commented:
I have only minor complaints.

derekwaynecarr commented:
Pushed update addressing comments:

1. Ensure we do not double wrap ForbiddenError
2. Removed debugging cruft
3. Moved to use api.ObjectMetaFor instead of Accessor
4. Added comment to note the issue with GenerateName happening after admission

lavalamp commented:
LGTM--Shippable seems to be two different flakes.
